src/polyline2_signed_distance_constraint.cpp

```diff
@@ -125,19 +128,6 @@ void Polyline2SignedDistanceConstraint::Quadraticize(const VectorXf& input,
         hess_coeff * unit_direction.x() * unit_direction.x();
     (*hess)(xidx_, yidx_) -= hess_xy;
     (*hess)(yidx_, xidx_) -= hess_xy;
-  } else {
-    // Closest point is a vertex.
-    const float grad_coeff = 2.0 * orientation * sign / level;
-    const float weighted_grad_coeff = weight_ * grad_coeff;
-    (*grad)(xidx_) += weighted_grad_coeff * dx;
-    (*grad)(yidx_) += weighted_grad_coeff * dy;
-
-    (*hess)(xidx_, xidx_) += weighted_grad_coeff * (grad_coeff * dx2 + 1.0);
-    (*hess)(yidx_, yidx_) += weighted_grad_coeff * (grad_coeff * dy2 + 1.0);
-
-    const float hess_xy = weighted_grad_coeff * grad_coeff * dx * dy;
-    (*hess)(xidx_, yidx_) += hess_xy;
```

Comment:
Wait why did this get deleted?

Reply:
ok I see this was an attempt to fix the case where closest point is a vertex. Unfortunately, this is going to break the case (which this was handling) where the closest point is indeed a vertex of the polyline but not an endpoint. To handle this correctly, I think you will need to edit the `ClosestPoint` function to set an `is_endpoint` flag when a returned closest point is actually an endpoint of the polyline.

Reply:
I put back all the cases where the closest point is a vertex. I implemented an if condition on the outside, first checking whether we are dealing with an endpoint.

Reply:
Ok no I see this now. But my other comment still stands - i.e. it looks like you're actually checking if the closest point is a vertex, not an endpoint.